### DISCUSSION TITLE
Implement decoding support for Opus-encoded .ogg files

### DISCUSSION THREAD

**Discussion by FibreTTP** [ppy/osu#25120] [Ideas]
As started by @pranshuthegamer in https://github.com/ppy/osu/issues/21757.

osu! has supported Vorbis audio in the .ogg container for a long time, but has lacked support for the more efficient [Opus codec](https://opus-codec.org) made by the same developers, and contained by the same container. It is an open format and beats all other audio codecs in audio efficiency, which would save space for everyone... for the few mappers that would put in the effort to encode their songs in it.

Discussion about raising the bitrate limit for MP3s (192kbps) has been [done before](https://osu.ppy.sh/community/forums/topics/1072965), and the conclusion is that "osu! shouldn't be [a] hq audio sharing platform." In addition, a developer has [stated](https://osu.ppy.sh/community/forums/posts/7120591) that "[higher quality audio] wont ever be accepted outside of FA tracks because [osu! is] not a lossless audio distribution service."

Therefore, implementation of this depends on an addition to the ranking criteria, namely: Should a beatmap wish to be ranked, it must not contain an opus-encoded file with an average bitrate greater than 96kbps. Unless its a featured artist's song, in which case it should be 192k.

**Comment by longnguyen2004** [comment]
> for the few mappers that would put in the effort to encode their songs in it.

That's not a problem. BNs are expected to do audio checks as part of the ranking process, so they're already familiar with the process of encoding. In addition, since we already have a clause in RC that disallow inflating the bitrate, we can also add a clause that encourages the use of modern codecs to lower file size. That way, BNs will have an incentive to encode to Opus.

Also, Opus only needs to go to 128kbps to be transparent, so we don't need to go all the way to 192kbps like Vorbis

**Reply by FibreTTP** [reply]
> 128kbps to be transparent

True. Though, out of the not many times I've tested, there was one case where 128k hasn't been enough -- and people smarter and with better hardware than me [have decided](https://wiki.hydrogenaud.io/index.php?title=Opus#Music_encoding_quality) that not even 192k is transparent.

Anyway, osu! is not an audio player, and the focus of the game is to hit circles (or bars or fruits or whatever), so I agree with lowering the bitrate limit. Why don't we just make it in-line with the non-FA songs at 96k for simplicity? (after all, I could only tell something was wrong when I was switching between bitrates)

And since you have more knowledge about mapping than me, do you suggest we _need_ to explicitly state **not** to encode files downloaded with [sketchy youtube audio downloader tool] into Opus? And that the Vorbis OGG file downloaded with [not sketchy youtube audio downloader tool] is perfectly fine for use in osu! natively?

**Reply by longnguyen2004** [reply]
> do you suggest we need to explicitly state not to encode files

I'd say that's up to the BN to decide, since they're already doing audio checks (see [this](https://osu.ppy.sh/beatmapsets/2086467/discussion/-/generalAll#/4048612), [this](https://osu.ppy.sh/beatmapsets/2021993/discussion/-/generalAll#/3922974/10553909), [this](https://osu.ppy.sh/beatmapsets/599960/discussion/-/generalAll#/3893160)), they can also provide a higher quality file if the existing one isn't suitable. For the mappers that do know about this stuff, a clause in the RC recommending the use of modern codecs should be enough I think, given that the majority of ranked maps are still using 192kbps mp3, probably due to being unaware that osu! supports Vorbis.

Disclaimer: I'm not a BN or a professional mapper, I'm just a nerd that sometimes follow the mapping scene. We should probably get the people that do this regularly for more opinion.

**Reply by longnguyen2004** [reply]
> sketchy youtube audio downloader tool

Anything other than youtube-dl/yt-dlp = sketchy to me :)

**Comment by cihe13375** [comment]
As a mapper myself I usually use mp3 instead of ogg because ogg (or maybe variable bitrate more accurately) is rumored to have some problems (at least one https://github.com/ppy/osu/issues/10671). Otherwise I would support this change.

**Comment by pranshuthegamer** [comment]
honestly, for some reason .opus files play weirdly on linux

**Reply by darkralts** [reply]
work fine on linux.

**Reply by darkralts** [reply]
make sure you have opus installed

**Comment by darkralts** [comment]
but yeah opus at 128kbps or 160kbps should sound almost or basically as good as a FLAC file, and lower bitrate than the current limits, makes sense if you ask me.

**Comment by tezvii** [comment]
the year is 2097. rhythm games have progressed to the point where people compete on who can think fast enough to hit hitobjects using a neural interface device.  osu!lazer still doesn't support opus.

**Comment by FibreTTP** [comment]
If any developers are reading this, what's blocking Opus support in osu! right now?

As far as I know, BASS supports Opus with a first-party add-on, and enabling it seems to be one line of code. Are there currently any differences in how the game handles MP3 and Vorbis streams that would make implementing Opus not as plug-and-play as I think it'd be (maybe something latency related)?

Or is it what I brought up in the OP, that Opus is too high quality for a circle-clicking game?

Sorry for "bumping" this, but it's been three years and we'd like a response. Thanks!
